### PR TITLE
HIVE-26809: Upgrade ORC to 1.8.1.

### PR DESCRIPTION
**Pull request by difin**
### What changes were proposed in this pull request?

Upgrading ORC version to currently latest version 1.8.1.
This PR is based on the changes proposed in uncompleted PR https://github.com/apache/hive/pull/2853 (ticket https://issues.apache.org/jira/browse/HIVE-25497 - Bump ORC to 1.7.2) with changes on top of it which enabled CI to pass.
Changes done in HIVE-25497:
"LLAP EncodedTreeReaderFactory is implementing its own TreeReaderFactory -- initially we are going to avoid LAZY IO here as everything is kept in memory."

Additional changes: Hive implements its own TreeReaderFactory. In ORC project, ORC-1060 - "Reduce memory usage when vectorized reading dictionary string encoding columns" introduced changes to StringDictionaryTreeReader which were causing exceptions in Hive EncodedTreeReaderFactory when attempting to upgrade to ORC 1.8.1. To handle that I added changes to Hive's EncodedTreeReaderFactory to use StringDictionaryTreeReader version as without ORC-1060. 

### Why are the changes needed?

To use latest ORC release in HIVE.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

CI tests passing.

### CODE REVIEW

**Comment by TuroczyX** [comment]
like it :)

**Review by cnauroth — CHANGES_REQUESTED** [review]
Hello @difin . Thank you for the patch. This looks like a good idea to try to complete before GA of Hive 4.0.

I see Apache ORC has just released version 1.8.1. Can we use that, so Hive gets on the latest release?

There are currently numerous test failures in CI, like this one:

http://ci.hive.apache.org/blue/organizations/jenkins/hive-precommit/detail/PR-3833/1/tests

I noticed a lot of `ArrayIndexOutOfBoundsException`, like this:

```
Caused by: java.lang.ArrayIndexOutOfBoundsException
	at java.lang.System.arraycopy(Native Method)
	at org.apache.orc.impl.TreeReaderFactory$StringDictionaryTreeReader.readDictionaryStream(TreeReaderFactory.java:2242)
	at org.apache.orc.impl.TreeReaderFactory$StringDictionaryTreeReader.nextVector(TreeReaderFactory.java:2283)
	at org.apache.orc.impl.TreeReaderFactory$StringTreeReader.nextVector(TreeReaderFactory.java:1963)
	at org.apache.hadoop.hive.ql.io.orc.encoded.EncodedTreeReaderFactory$StringStreamReader.nextVector(EncodedTreeReaderFactory.java:313)
	at org.apache.hadoop.hive.llap.io.decode.OrcEncodedDataConsumer.decodeBatch(OrcEncodedDataConsumer.java:196)
	at org.apache.hadoop.hive.llap.io.decode.OrcEncodedDataConsumer.decodeBatch(OrcEncodedDataConsumer.java:66)
	at org.apache.hadoop.hive.llap.io.decode.EncodedDataConsumer.consumeData(EncodedDataConsumer.java:122)
	at org.apache.hadoop.hive.llap.io.encoded.SerDeEncodedDataReader.sendEcbToConsumer(SerDeEncodedDataReader.java:1687)
	at org.apache.hadoop.hive.llap.io.encoded.SerDeEncodedDataReader.processOneSlice(SerDeEncodedDataReader.java:1059)
	at org.apache.hadoop.hive.llap.io.encoded.SerDeEncodedDataReader.processOneFileSplit(SerDeEncodedDataReader.java:908)
	at org.apache.hadoop.hive.llap.io.encoded.SerDeEncodedDataReader.readFileWithCache(SerDeEncodedDataReader.java:859)
	at org.apache.hadoop.hive.llap.io.encoded.SerDeEncodedDataReader.performDataRead(SerDeEncodedDataReader.java:731)
	at org.apache.hadoop.hive.llap.io.encoded.SerDeEncodedDataReader$5.run(SerDeEncodedDataReader.java:278)
	at org.apache.hadoop.hive.llap.io.encoded.SerDeEncodedDataReader$5.run(SerDeEncodedDataReader.java:275)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:422)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1878)
	at org.apache.hadoop.hive.llap.io.encoded.SerDeEncodedDataReader.callInternal(SerDeEncodedDataReader.java:275)
	at org.apache.hadoop.hive.llap.io.encoded.SerDeEncodedDataReader.callInternal(SerDeEncodedDataReader.java:115)
	at org.apache.tez.common.CallableWithNdc.call(CallableWithNdc.java:36)
	at org.apache.hadoop.hive.llap.io.decode.EncodedDataConsumer$CpuRecordingCallable.call(EncodedDataConsumer.java:88)
	at org.apache.hadoop.hive.llap.io.decode.EncodedDataConsumer$CpuRecordingCallable.call(EncodedDataConsumer.java:73)
	... 5 more
```

Can you please investigate?

**Comment by difin** [comment]
> Hello @difin . Thank you for the patch. This looks like a good idea to try to complete before GA of Hive 4.0.
> 
> I see Apache ORC has just released version 1.8.1. Can we use that, so Hive gets on the latest release?
> 
> There are currently numerous test failures in CI, like this one:
> 
> http://ci.hive.apache.org/blue/organizations/jenkins/hive-precommit/detail/PR-3833/1/tests
> 
> I noticed a lot of `ArrayIndexOutOfBoundsException`, like this:
> 
> ```
> Caused by: java.lang.ArrayIndexOutOfBoundsException
> 	at java.lang.System.arraycopy(Native Method)
> 	at org.apache.orc.impl.TreeReaderFactory$StringDictionaryTreeReader.readDictionaryStream(TreeReaderFactory.java:2242)
> 	at org.apache.orc.impl.TreeReaderFactory$StringDictionaryTreeReader.nextVector(TreeReaderFactory.java:2283)
> 	at org.apache.orc.impl.TreeReaderFactory$StringTreeReader.nextVector(TreeReaderFactory.java:1963)
> 	at org.apache.hadoop.hive.ql.io.orc.encoded.EncodedTreeReaderFactory$StringStreamReader.nextVector(EncodedTreeReaderFactory.java:313)
> 	at org.apache.hadoop.hive.llap.io.decode.OrcEncodedDataConsumer.decodeBatch(OrcEncodedDataConsumer.java:196)
> 	at org.apache.hadoop.hive.llap.io.decode.OrcEncodedDataConsumer.decodeBatch(OrcEncodedDataConsumer.java:66)
> 	at org.apache.hadoop.hive.llap.io.decode.EncodedDataConsumer.consumeData(EncodedDataConsumer.java:122)
> 	at org.apache.hadoop.hive.llap.io.encoded.SerDeEncodedDataReader.sendEcbToConsumer(SerDeEncodedDataReader.java:1687)
> 	at org.apache.hadoop.hive.llap.io.encoded.SerDeEncodedDataReader.processOneSlice(SerDeEncodedDataReader.java:1059)
> 	at org.apache.hadoop.hive.llap.io.encoded.SerDeEncodedDataReader.processOneFileSplit(SerDeEncodedDataReader.java:908)
> 	at org.apache.hadoop.hive.llap.io.encoded.SerDeEncodedDataReader.readFileWithCache(SerDeEncodedDataReader.java:859)
> 	at org.apache.hadoop.hive.llap.io.encoded.SerDeEncodedDataReader.performDataRead(SerDeEncodedDataReader.java:731)
> 	at org.apache.hadoop.hive.llap.io.encoded.SerDeEncodedDataReader$5.run(SerDeEncodedDataReader.java:278)
> 	at org.apache.hadoop.hive.llap.io.encoded.SerDeEncodedDataReader$5.run(SerDeEncodedDataReader.java:275)
> 	at java.security.AccessController.doPrivileged(Native Method)
> 	at javax.security.auth.Subject.doAs(Subject.java:422)
> 	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1878)
> 	at org.apache.hadoop.hive.llap.io.encoded.SerDeEncodedDataReader.callInternal(SerDeEncodedDataReader.java:275)
> 	at org.apache.hadoop.hive.llap.io.encoded.SerDeEncodedDataReader.callInternal(SerDeEncodedDataReader.java:115)
> 	at org.apache.tez.common.CallableWithNdc.call(CallableWithNdc.java:36)
> 	at org.apache.hadoop.hive.llap.io.decode.EncodedDataConsumer$CpuRecordingCallable.call(EncodedDataConsumer.java:88)
> 	at org.apache.hadoop.hive.llap.io.decode.EncodedDataConsumer$CpuRecordingCallable.call(EncodedDataConsumer.java:73)
> 	... 5 more
> ```
> 
> Can you please investigate?

Hi @cnauroth, thank you for your comments. I am investigating the CI errors and will upgrade to ORC 1.8.1 too.

File: ql/src/java/org/apache/hadoop/hive/ql/io/orc/encoded/EncodedTreeReaderFactory.java [hunk 1]
```diff
@@ -224,7 +237,252 @@ private static void skipCompressedIndex(boolean isCompressed, PositionProvider i
     index.getNext();
   }
 
-  protected static class StringStreamReader extends StringTreeReader
+  public static class StringDictionaryTreeReaderHive extends TreeReader {
```
Comment (ayushtkn) [inline ql/src/java/org/apache/hadoop/hive/ql/io/orc/encoded/EncodedTreeReaderFactory.java:240]: Why is this added in scope of upgrade?

File: ql/src/java/org/apache/hadoop/hive/ql/io/orc/encoded/EncodedTreeReaderFactory.java [hunk 2]
```diff
@@ -1249,10 +1534,13 @@ public void seek(PositionProvider index) throws IOException {
     }
 
     @Override
-    public void nextVector(
-      ColumnVector previousVector, boolean[] isNull, int batchSize) throws IOException {
+    public void nextVector(ColumnVector previousVector,
+        boolean[] isNull,
+        final int batchSize,
+        FilterContext filterContext,
+        ReadPhase readPhase) throws IOException {
```
Comment (ayushtkn) [inline ql/src/java/org/apache/hadoop/hive/ql/io/orc/encoded/EncodedTreeReaderFactory.java:1541]: this code doesn't look formatted

**Review by ayushtkn — COMMENTED** [review]
minor comments and a question why are these new classes added?

File: llap-server/src/java/org/apache/hadoop/hive/llap/io/decode/OrcEncodedDataConsumer.java [hunk 3]
```diff
@@ -61,7 +61,7 @@
 import org.apache.orc.impl.TreeReaderFactory.TreeReader;
 import org.apache.orc.impl.WriterImpl;
 import org.apache.orc.OrcProto;
-
+import org.apache.orc.impl.reader.tree.TypeReader;
```
[the remark this answers is not in the document]
Comment (difin) [reply llap-server/src/java/org/apache/hadoop/hive/llap/io/decode/OrcEncodedDataConsumer.java]: fixed

File: ql/src/java/org/apache/hadoop/hive/ql/io/orc/encoded/EncodedTreeReaderFactory.java (hunk 1 above)
Comment (ayushtkn) [reply ql/src/java/org/apache/hadoop/hive/ql/io/orc/encoded/EncodedTreeReaderFactory.java:237]: Ok, that seems to be an improvement or say a bug fix in the ORC project and we are just implementing our own varient because now the original class is causing test failures.
This isn't the ideal approach and will backfire in future when we try to upgrade and the changes in ORC depends on the ones which we ditched.

We should try to adapt to those changes and make sure we don't crash with those changes in Hive by making hive changes, rather than maintaining a old version of ORC class at Hive

Comment (difin) [reply ql/src/java/org/apache/hadoop/hive/ql/io/orc/encoded/EncodedTreeReaderFactory.java:237]: Hi @ayushtkn, I agree with you. It is not ideal approach. Before implementing this approach I did try to adapt Hive, but I didn't succeed to find how Hive could be adapted to ORC-1060 changes because those changes are only inside internal implementation of Orc StringDictionaryTreeReader class. 

I agree with you that this approach is not ideal and will backfire in future when we try to upgrade and the changes in ORC depends on the ones which we ditched, but Hive already heavily depends on internal ORC API by implementing its own column readers on top of ORC and when upgrading to different ORC version it is often required to make adaptations in Hive.

Comment (ayushtkn) [reply ql/src/java/org/apache/hadoop/hive/ql/io/orc/encoded/EncodedTreeReaderFactory.java:237]: @abstractdog any pointers here? Does that sound fine thing to do for now

Comment (ayushtkn) [reply ql/src/java/org/apache/hadoop/hive/ql/io/orc/encoded/EncodedTreeReaderFactory.java:237]: makes sense to me, thanx Laszlo!!!
